{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "5b2c7c2d-cd42-4e2f-b947-c59799ec3967",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Protein ID: gi|25091097|sp|O60216.2|RAD21_HUMAN, Sequence Length: 631\n",
      "Protein ID: gi|341942169|sp|Q61550.3|RAD21_MOUSE, Sequence Length: 635\n",
      "Protein ID: gi|82187382|sp|Q6TEL1.1|RA21A_DANRE, Sequence Length: 643\n",
      "Protein ID: gi|29336593|sp|O93310.1|RAD21_XENLA, Sequence Length: 629\n",
      "Protein ID: gi|110287797|sp|Q3SWX9.1|RAD21_BOVIN, Sequence Length: 630\n",
      "Protein ID: gi|327478531|sp|A2AU37.2|RD21L_MOUSE, Sequence Length: 552\n",
      "Protein ID: gi|259016327|sp|Q9H4I0.3|RD21L_HUMAN, Sequence Length: 556\n",
      "Protein ID: gi|327488389|sp|D2HSB3.1|RD21L_AILME, Sequence Length: 554\n"
     ]
    }
   ],
   "source": [
    "from Bio import SeqIO\n",
    "\n",
    "# File path\n",
    "fasta_file = \"/Users/mohanavenkataphaneendrareddyalla/Downloads/RAD21_Protein_Sequences.fasta\"\n",
    "\n",
    "# Parse the fasta file and print each protein ID and its sequence length\n",
    "for record in SeqIO.parse(fasta_file, \"fasta\"):\n",
    "    print(f\"Protein ID: {record.id}, Sequence Length: {len(record.seq)}\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "698f0cc4-ce3c-4557-acc8-5a0698f59aa2",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "\n",
      "Building a new DB, current time: 11/18/2024 01:45:18\n",
      "New DB name:   /Users/mohanavenkataphaneendrareddyalla/RAD21_Search_Database\n",
      "New DB title:  /Users/mohanavenkataphaneendrareddyalla/Downloads/RAD21_Protein_Sequences.fasta\n",
      "Sequence type: Protein\n",
      "Deleted existing Protein BLAST database named /Users/mohanavenkataphaneendrareddyalla/RAD21_Search_Database\n",
      "Keep MBits: T\n",
      "Maximum file size: 3000000000B\n",
      "Adding sequences from FASTA; added 8 sequences in 0.0075469 seconds.\n",
      "\n",
      "\n",
      "Database created successfully.\n"
     ]
    }
   ],
   "source": [
    "import subprocess\n",
    "\n",
    "# Path for the BLAST database to be created\n",
    "database_name = \"/Users/mohanavenkataphaneendrareddyalla/RAD21_Search_Database\"\n",
    "\n",
    "try:\n",
    "    subprocess.run([\"makeblastdb\", \"-in\", fasta_file, \"-dbtype\", \"prot\", \"-out\", database_name], check=True)\n",
    "    print(\"Database created successfully.\")\n",
    "except subprocess.CalledProcessError as e:\n",
    "    print(f\"Error creating the database: {e}\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "bdbdb811-7ad0-4404-acb1-b43897a1375f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "BLASTp executed successfully.\n"
     ]
    }
   ],
   "source": [
    "try:\n",
    "    subprocess.run([\"blastp\", \"-query\", fasta_file, \"-db\", database_name, \"-outfmt\", \"5\", \"-out\", \"RAD21_Blast_Result.out\"], check=True)\n",
    "    print(\"BLASTp executed successfully.\")\n",
    "except subprocess.CalledProcessError as e:\n",
    "    # Catching errors if the 'blastp' command fails\n",
    "    print(f\"Error running BLASTp: {e}\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "2bca2ae8-2f94-44ea-9dce-ed8dbca9bd1c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Debug - Listing all queries in the file:\n",
      "Query: gi|25091097|sp|O60216.2|RAD21_HUMAN RecName: Full=Double-strand-break repair protein rad21 homolog; Short=hHR21; AltName: Full=Nuclear matrix protein 1; Short=NXP-1; AltName: Full=SCC1 homolog; Contains: RecName: Full=64-kDa C-terminal product; AltName: Full=64-kDa carboxy-terminal product; AltName: Full=65-kDa carboxy-terminal product\n",
      "Query: gi|341942169|sp|Q61550.3|RAD21_MOUSE RecName: Full=Double-strand-break repair protein rad21 homolog; Short=mHR21; AltName: Full=Pokeweed agglutinin-binding protein 29; Short=PW29; AltName: Full=SCC1 homolog; Contains: RecName: Full=64-kDa C-terminal product; AltName: Full=64-kDa carboxy-terminal product\n",
      "Query: gi|82187382|sp|Q6TEL1.1|RA21A_DANRE RecName: Full=Double-strand-break repair protein rad21 homolog A; AltName: Full=SCC1 homolog; Contains: RecName: Full=64-kDa C-terminal product; AltName: Full=64-kDa carboxy-terminal product\n",
      "Query: gi|29336593|sp|O93310.1|RAD21_XENLA RecName: Full=Double-strand-break repair protein rad21 homolog; AltName: Full=SCC1 homolog; Contains: RecName: Full=64-kDa C-terminal product; AltName: Full=64-kDa carboxy-terminal product\n",
      "Query: gi|110287797|sp|Q3SWX9.1|RAD21_BOVIN RecName: Full=Double-strand-break repair protein rad21 homolog; Contains: RecName: Full=64-kDa C-terminal product; AltName: Full=64-kDa carboxy-terminal product\n",
      "Query: gi|327478531|sp|A2AU37.2|RD21L_MOUSE RecName: Full=Double-strand-break repair protein rad21-like protein 1\n",
      "Query: gi|259016327|sp|Q9H4I0.3|RD21L_HUMAN RecName: Full=Double-strand-break repair protein rad21-like protein 1\n",
      "Query: gi|327488389|sp|D2HSB3.1|RD21L_AILME RecName: Full=Double-strand-break repair protein rad21-like protein 1\n",
      "\n",
      "Debug - Checking hits and E-values for RAD21_HUMAN:\n",
      "\n",
      "Hits for gi|25091097|sp|O60216.2|RAD21_HUMAN RecName: Full=Double-strand-break repair protein rad21 homolog; Short=hHR21; AltName: Full=Nuclear matrix protein 1; Short=NXP-1; AltName: Full=SCC1 homolog; Contains: RecName: Full=64-kDa C-terminal product; AltName: Full=64-kDa carboxy-terminal product; AltName: Full=65-kDa carboxy-terminal product:\n",
      "E-value: 0.0\n",
      "E-value: 0.0\n",
      "E-value: 0.0\n",
      "E-value: 0.0\n",
      "E-value: 0.0\n",
      "E-value: 6.4792e-150\n",
      "E-value: 6.55153e-144\n",
      "E-value: 9.27111e-144\n"
     ]
    }
   ],
   "source": [
    "from Bio.Blast import NCBIXML\n",
    "\n",
    "with open(\"/Users/mohanavenkataphaneendrareddyalla/RAD21_Blast_Result.out\") as result_handle:\n",
    "    blast_records = NCBIXML.parse(result_handle)\n",
    "    \n",
    "  \n",
    "    print(\"Debug - Listing all queries in the file:\")\n",
    "    for blast_record in blast_records:\n",
    "        print(f\"Query: {blast_record.query}\")\n",
    "    \n",
    "   \n",
    "    result_handle.seek(0)\n",
    "    blast_records = NCBIXML.parse(result_handle)\n",
    "    \n",
    "    print(\"\\nDebug - Checking hits and E-values for RAD21_HUMAN:\")\n",
    "    for blast_record in blast_records:\n",
    "        if \"RAD21_HUMAN\" in blast_record.query:\n",
    "            print(f\"\\nHits for {blast_record.query}:\")\n",
    "            if len(blast_record.alignments) > 0:\n",
    "                for alignment in blast_record.alignments:\n",
    "                    for hsp in alignment.hsps:\n",
    "                        \n",
    "                        print(f\"E-value: {hsp.expect}\")\n",
    "            else:\n",
    "                print(f\"No hits found for {blast_record.query}\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9dd3bfae-264b-4cc5-9817-8128b8078f0e",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
